File: 24-Q01/q1.rb
#!/usr/bin/env ruby

require 'json'

f = File.open('./q1-input.txt')

l1 = []
l2 = []
f.readlines.each do |line|
  e1, e2 = line.split(' ')

  sortedIndex1 = 0
  l1.each do |x|
    if x > e1 then break end
    sortedIndex1 += 1
  end
  l1.insert(sortedIndex1, e1)

  sortedIndex2 = 0
  l2.each do |x|
    if x > e2 then break end
    sortedIndex2 += 1
  end
  l2.insert(sortedIndex2, e2)
end

p l1
p l2

distances = l1.map.with_index do |e1, i| 
  e2 = l2[i]

  tmp = (e1.to_i - e2.to_i).abs
end
p distances

p distances.sum

File: 24-Q19/q1.py
#!/usr/bin/env python

patterns = []
tasks = []

with open("input.txt", "r") as file:
	lines = file.readlines()

	patterns = lines[0].split(',')
	patterns = [p.strip() for p in patterns]

	tasks = [line.strip() for line in lines[2:]]

print(patterns)
print(tasks)

cache = {}
def can_be_arranged(task, offset):
	# print('can_be_arranged({}, {})'.format(task, offset))
	if (task, offset) in cache:
		return cache[(task, offset)]
	i = 0
	for pattern in patterns:
		i += 1
		if offset == 0:
			print('\tTesting pattern {} ({}) of {}'.format(i+1, pattern, len(patterns)))
		if task[offset:] == pattern:
			return True

		if task[offset:offset + len(pattern)] == pattern:
			if can_be_arranged(task, offset + len(pattern)):
				cache[(task, offset)] = True
				return True

	cache[(task, offset)] = False
	return False

sum = 0
for task in tasks:
	if can_be_arranged(task, 0):
		sum += 1

print("Sum: {}".format(sum))
